Include cleanup

MaxKellermann opened the pull request:
Two commits which acculumate to a 20% build speedup.

MaxKellermann commented:
There were CI build failures because lots of includes were still missing. These were always missing, and hose bugs only now become visible with stricter includes. Will fix and force-push until all CI targets succeed the build.

MaxKellermann commented:
Build time master / Fedora 30: 10m 04s
Build time this PR / Fedora 30: 5m 36s
(MacOS still doesn't build, but I'll do another force-push now.)

MaxKellermann commented:
All CI tests succeed now (plus two false negatives). Phew.

MaxKellermann commented:
Rebased and fixed merge conflict in `iosource/Manager.cc`.

timwoj commented on src/Anon.cc:
```diff
@@ -1,14 +1,14 @@
 #include "Val.h"
 #include "NetVar.h"
 
+#include <stdlib.h>
```
The style guide specifies that system headers should come before all other headers (see https://github.com/zeek/zeek-docs/blob/master/devel/contributors/coding-style.rst#including-files). I'm inclined to adjust that to let the class header come before them. Is there another reason for moving the system headers to the end?

MaxKellermann replied:
`Foo.cc` should always include `Foo.h` first to validate that `Foo.h` is self-containing and can be included by others without preconditions.
From there, it seems natural to me that "custom first, system last" is the natural order. We know/assert that system headers must be correct, so we include them in the order they need to be validated. But that's just a subjective opinion, and if it violates your coding style, I'll change it.

timwoj replied:
The style is more along the lines of include-what-you-use, meaning if the source file needs it, it shouldn't be relying on some other header to include it. The easiest way to ensure that is the put them first. If some other header includes it, then include guards will prevent the double-include. Does that make enough sense?

I'll make a note to update the style document to put including the class header first.

timwoj commented on src/Anon.h:
```diff
@@ -11,11 +11,11 @@
 #include "Reporter.h"
-#include "net_util.h"
+
+using std::map;
```
We generally try to avoid `using` statements in headers, if possible, especially for `std` types. We had a PR recently that went through and removed a bunch of uses of `using namespace std` from all of the headers.

MaxKellermann replied:
Agree, I'd never do that - but where I added `using` statements, it was necessary to make the code build, because those sources depended on `using` statements in other headers which are no longer included.
Removing those `using` statements and adding namespace prefixes would be the next step.
But for this PR, I wanted to keep the code untouched as much as possible, and only change the includes.

timwoj replied:
Ok, that makes sense if you're going to follow-up and remove them later. 👍

timwoj commented on src/analyzer/protocol/tcp/TCP_Reassembler.h:
```diff
@@ -1,13 +1,18 @@
-namespace analyzer { namespace tcp {
+namespace analyzer {
+
+class Analyzer;
```
Isn't this forward-declared in `analyzer/Analyzer.h` which you added?

MaxKellermann replied:
Right, first I added the include to make it build, then I decided I can replace it with a forward declaration, but then forgot to remove the include... fixed now.

timwoj commented on src/logging/Tag.h:
```diff
@@ -3,13 +3,17 @@
+template <class T, class C>
+class ComponentManager;
+}
+
```
Redefining this block in multiple places (see also `input/Tag.h`) feels a little brittle, not that it's likely to ever change.

MaxKellermann replied:
Each header which needs a certain symbol should have a copy of the forward declaration, even if another header has already done it. Imagine that other header one day doesn't need the declaration anymore, or that header isn't included anymore - then `logging/Tag.h` must not break, and should not need to be adjusted. It should be self-contained.
Having multiple forward declarations for one symbol is allowed in C++.

jsiwek commented:
> * Do we want to standardize on where `zeek-config.h` gets included in the order of headers? In most places it's the first thing since it defines a bunch of configure-time options, but there's other places where that's not true.

Not something I'd probably remember to be too strict about, but yeah, my general expectation is that gets included "first, or as early as possible" such that it will already be in the right place if you need to use its definitions in subsequence platform-specific header inclusion logic.  So yes, worth suggesting that order as a standard practice.

> * Do we want to switch over to using the new "standard library" version of the C headers (cstdint, cmath, etc)?

I'd favor that, but not hugely significant IMO.  Also can be done independently/whenever.

MaxKellermann commented on src/Event.cc:
```diff
@@ -102,6 +103,26 @@ EventMgr::~EventMgr()
+		}
+	}
+
+void EventMgr::Dispatch(Event* event, bool no_remote)
```
Will do.

MaxKellermann replied:
There's a problem with that: right now, the order in `Queue.cc` doesn't match the header already.
Header:
- `~EventMgr()`
- `Drain(...)`
- `QueueEvent(Event*)`
Source:
- `~EventMgr()`
- `QueueEvent(Event*)`
- `Drain()`
But anyway, I've moved it as you requested.

MaxKellermann commented:
>     * Do we want to standardize on where `zeek-config.h` gets included in the order of headers?

gcc and clang have the (non-standard) option `-include`, which the Linux kernel uses, for example. Zeek could decide to depend on that feature and thus force `zeek-config.h` to be always included in each compilation unit, without requiring this policy.
If you decide not to use that option, there should be a CI test which verifies that all (relevant) sources include this header first. Or else there may be interesting runtime breakages. This is dangerous and hard to debug.

>     * Do we want to switch over to using the new "standard library" version of the C headers (cstdint, cmath, etc)?

I don't have an opinion on this, but I think there's a `clang-tidy` option to do this automatically. If you decide to do so, I can give you a tree-wide PR for it.

timwoj commented:
> gcc and clang have the (non-standard) option -include, which the Linux kernel uses, for example. Zeek could decide to depend on that feature and thus force zeek-config.h to be always included in each compilation unit, without requiring this policy.

That sounds totally fine to me.

> I'd favor that, but not hugely significant IMO. Also can be done independently/whenever.

I think the only concern is that it moves all of the methods into the std namespace, which will cause a lot of build breakage until they get sorted. It could lead to breakage for external plugins and whatnot too.

MaxKellermann commented:
> I think the only concern is that it moves all of the methods into the std namespace, which will cause a lot of build breakage until they get sorted.

True, Zeek would then need to refer to those symbols with `std::` qualification. Though in practice, the plain C symbols will be available as well in all C++ implementations I know, so I don't think anything will really (visibly) break. But in any case, for correctness, we should really use `std::`, but there's no immediate hurry.

> It could lead to breakage for external plugins and whatnot too.

If this breaks a plugin, then the plugin was already broken before, it just wasn't visible... having stricter semantics is a good thing, even if this reveals existing old breakages. Revealing mistakes means mistakes get fixed. That's good, even if it means some short-term annoyance.

MaxKellermann commented:
Force-pushed a new revision:
- rebased on master
- reverted the system include order according to the coding style documentation, with one exception: a file's own header is still included first
- removed the unnecessary new include from `TCP_Reassembler.h`

timwoj commented on src/DNS_Mgr.h:
```diff
@@ -2,18 +2,16 @@
+#include <map>
+#include <queue>
+#include <utility>
+
```
This block needs to move up

timwoj commented on src/util.cc:
```diff
@@ -42,6 +42,7 @@
+#include "Desc.h"
 #include "digest.h"
 #include "input.h"
 #include "util.h"
```
Move "util.h" up?

timwoj commented on src/probabilistic/Hasher.cc:
```diff
@@ -1,13 +1,16 @@
 
+#include <broker/data.hh>
+
+#include <openssl/evp.h>
```
Move this one and the one below it up in the order.

timwoj commented on src/plugin/Plugin.h:
```diff
@@ -2,23 +2,21 @@
 #pragma once
 
+#include "zeek-config.h"
+#include "logging/WriterBackend.h"
```
Move this one down.

timwoj commented on src/plugin/ComponentManager.h:
```diff
@@ -1,15 +1,15 @@
 #include "zeekygen/Manager.h"
+#include "DebugLogger.h"
+
+#include <map>
```
Move these up

timwoj commented on src/iosource/PktSrc.h:
```diff
@@ -2,13 +2,16 @@
-#include "Dict.h"
 #include "Packet.h"
 
+#include <vector>
```
Move this one and the one below it up.

timwoj commented on src/iosource/Manager.h:
```diff
@@ -4,12 +4,15 @@
 
 #include "zeek-config.h"
 
+#include "IOSource.h"
```
Undo the change here to move these ones up.

timwoj commented on src/file_analysis/File.h:
```diff
@@ -2,21 +2,22 @@
 #include "BroString.h"
 #include "WeirdState.h"
 
+#include <string>
```
Move these up

timwoj commented on src/broker/Manager.h:
```diff
@@ -1,5 +1,9 @@
 #pragma once
 
+#include "NetVar.h"
```
The internal headers and system headers need to be swapped in this one.

timwoj commented on src/analyzer/protocol/icmp/ICMP.cc:
```diff
@@ -1,17 +1,20 @@
 
 #include "events.bif.h"
 
+#include <algorithm>
```
Move this up

timwoj commented on src/analyzer/protocol/login/NVT.cc:
```diff
@@ -2,15 +2,16 @@
 
 #include "events.bif.h"
 
+#include <stdlib.h>
```
Move this up

timwoj commented on src/analyzer/protocol/mime/MIME.h:
```diff
@@ -1,15 +1,18 @@
 #pragma once
 
+#include "BroString.h"
```
Revert reordering  this one

timwoj commented on src/analyzer/protocol/stepping-stone/SteppingStone.cc:
```diff
@@ -1,18 +1,17 @@
-
 #include "events.bif.h"
 
+#include <stdlib.h>
```
Move this up.

MaxKellermann commented:
Rebased and fixed after @timwoj's comments
